File: campeones_mundial.R
rm(list = ls())

library(dplyr)
library(magrittr)
library(ggplot2)
extrafont::loadfonts(device = "win")

# Import Data
partidos_fifa_copa_mundial_procesado <-
  readr::read_delim(
    "https://raw.githubusercontent.com/cienciadedatos/datos-de-miercoles/master/datos/2019/2019-04-10/partidos.txt",
    delim = "\t"
  )

# Encontrar campeón y año respectivo

campeones_mundial <- partidos_fifa_copa_mundial_procesado %>%
  mutate(partido_orden = as.numeric(stringr::str_extract(partido_orden, "[[:digit:]]+"))) %>% # Dejar sólo los números
  group_by(anio) %>%
  mutate(
    ganador_partido = ifelse(equipo_1_final >= equipo_2_final , equipo_1, equipo_2),
    ganador_goles = ifelse(
      equipo_1_final >= equipo_2_final ,
      equipo_1_final  ,
      equipo_2_final
    ),
    ganador_mundial = ganador_partido[n()]
  ) %>%
  filter(ganador_partido == ganador_mundial) %>%
  ungroup() %>%
  group_by(anio, ganador_mundial) %>%
  summarise(goles = sum(ganador_goles))

# Gráfico

windowsFonts(Roboto = windowsFont("Roboto Medium"))

ggplot(data = campeones_mundial, aes(x = anio, y = goles, label = ganador_mundial)) +
  geom_point() +
  geom_line() +
  geom_segment(
    aes(
      x = min(anio) - 1,
      xend = max(anio) + 1,
      y = mean(campeones_mundial$goles),
      yend = mean(campeones_mundial$goles)
    ),
    linetype = "dashed",
    color = "red"
  ) +
  annotate(
    "text",
    x = 2018,
    y = mean(campeones_mundial$goles),
    label = round(mean(campeones_mundial$goles), 1),
    hjust = -0.55,
    size = 5,
    color = "red",
    fontface = 2
  ) +
  labs(
    x = "Año",
    y = "Número de Goles",
    title = "Distribución de Goles del país ganador del Mundial",
    subtitle = "Por mundial, el equipo ganador marca aproximadamente unos 15 goles",
    caption = "Fuente: Datos De Miércoles"
  ) +
  theme_minimal() +
  theme(text = element_text(size = 16, family = "Roboto Medium")) +
  scale_x_continuous(
    limits = c(min(campeones_mundial$anio) - 1, y = max(campeones_mundial$anio) +
                 5),
    expand = c(0, 2),
    breaks = seq(
      min(campeones_mundial$anio),
      max(campeones_mundial$anio) + 2,
      by = 4
    )
  )

ggsave(
  "campeones_gol.png",
  dpi = 600,
  height = 6,
  width = 12
)

File: campeones_mundial.py
import pandas as pd
import matplotlib.pyplot as plt
from numpy import round

# leer data
partidos_fifa = pd.read_csv(
    "https://raw.githubusercontent.com/cienciadedatos/datos-de-miercoles/master/datos/2019/2019-04-10/partidos.txt", sep="\t")

# Dejar sólo dígitos
partidos_fifa.partido_orden = partidos_fifa.partido_orden.str.extract('(\d+)')

# Crear columna con ganador partido
partidos_fifa["ganador_partido"] = partidos_fifa.apply(
    lambda x: x['equipo_1'] if x['equipo_1_final'] >= x['equipo_2_final']
    else x['equipo_2'], axis=1)

# Crear columna con goles del ganador
partidos_fifa["ganador_goles"] = partidos_fifa.apply(
    lambda x: x['equipo_1_final'] if x['equipo_1_final'] >= x['equipo_2_final']
    else x['equipo_2_final'], axis=1)

# Agrupar por año y hallar los últimos
campeones_fifa = partidos_fifa.groupby("anio").last()
campeones_fifa = campeones_fifa.loc[:, ["ganador_partido"]]
campeones_fifa.rename(
    columns={"ganador_partido": "ganador_mundial"}, inplace=True)

# Left join
partidos_fifa = pd.merge(partidos_fifa, campeones_fifa, how="left", on="anio")
partidos_fifa = partidos_fifa[partidos_fifa["ganador_partido"]
                              == partidos_fifa["ganador_mundial"]]
campeones_goles = partidos_fifa.loc[:, [
    'anio', 'ganador_goles', 'ganador_mundial']]
# Sumar goles
campeones_goles = campeones_goles.groupby(["anio", "ganador_mundial"]).sum()

# Gráfico
plt.style.use(['ggplot'])
plt.scatter(campeones_goles.index.get_level_values(
    'anio'), campeones_goles['ganador_goles'], color='k')
plt.plot(campeones_goles.index.get_level_values('anio'),
         campeones_goles['ganador_goles'], linestyle='solid', color='k')
plt.axhline(
    y=round(campeones_goles["ganador_goles"].mean(), 1), linestyle="--", color='r')
plt.show()
